File: TAREA1/3_bucles_anidados.cpp
#include <iostream>
#include <chrono>
#include <cstring>
#include <cstdlib>
#include <iomanip>
#include <vector>
#include <string>

// ─── Tamaños a evaluar ───────────────────────────────────────────
static const int SIZES[] = {64, 128, 256, 512, 1024};
static const int NSIZES  = 5;

// ─── Matrices estáticas máximas (evita malloc en el benchmark) ───
static const int MAXN = 1024;
static double A[MAXN][MAXN], B[MAXN][MAXN], C[MAXN][MAXN];

// ─── Inicialización ──────────────────────────────────────────────
void init(int N) {
    srand(42);
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++) {
            A[i][j] = (double)rand() / RAND_MAX;
            B[i][j] = (double)rand() / RAND_MAX;
        }
}

void zero_C(int N) {
    for (int i = 0; i < N; i++)
        memset(C[i], 0, N * sizeof(double));
}

// ─── Los 6 órdenes de bucle ──────────────────────────────────────
// Mismo resultado matemático, distinto patrón de acceso a caché.

// ijk: el clásico "natural". B[k][j] tiene stride-N (columnar) → malo
void matmul_ijk(int N) {
    zero_C(N);
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            for (int k = 0; k < N; k++)
                C[i][j] += A[i][k] * B[k][j];
}

// ikj: el mejor orden. j varía en el inner loop → A[i][k] escalar,
//       B[k][j] y C[i][j] ambos stride-1 ✓
void matmul_ikj(int N) {
    zero_C(N);
    for (int i = 0; i < N; i++)
        for (int k = 0; k < N; k++) {
            double r = A[i][k];          // escalar: sólo 1 carga
            for (int j = 0; j < N; j++)
                C[i][j] += r * B[k][j]; // ambas matrices stride-1
        }
}

// jik
void matmul_jik(int N) {
    zero_C(N);
    for (int j = 0; j < N; j++)
        for (int i = 0; i < N; i++)
            for (int k = 0; k < N; k++)
                C[i][j] += A[i][k] * B[k][j];
}

// jki
void matmul_jki(int N) {
    zero_C(N);
    for (int j = 0; j < N; j++)
        for (int k = 0; k < N; k++) {
            double r = B[k][j];
            for (int i = 0; i < N; i++)
                C[i][j] += A[i][k] * r;
}
}

// kij
void matmul_kij(int N) {
    zero_C(N);
    for (int k = 0; k < N; k++)
        for (int i = 0; i < N; i++) {
            double r = A[i][k];
            for (int j = 0; j < N; j++)
                C[i][j] += r * B[k][j];
        }
}

// kji
void matmul_kji(int N) {
    zero_C(N);
    for (int k = 0; k < N; k++)
        for (int j = 0; j < N; j++) {
            double r = B[k][j];
            for (int i = 0; i < N; i++)
                C[i][j] += A[i][k] * r;
        }
}

// ─── Benchmark ───────────────────────────────────────────────────
typedef void (*MatmulFn)(int);

double measure(MatmulFn fn, int N, int runs = 3) {
    double total = 0;
    for (int r = 0; r < runs; r++) {
        auto t0 = std::chrono::high_resolution_clock::now();
        fn(N);
        auto t1 = std::chrono::high_resolution_clock::now();
        total += std::chrono::duration<double, std::milli>(t1 - t0).count();
    }
    return total / runs;
}

int main() {
    struct Order { const char* name; MatmulFn fn; };
    Order orders[] = {
        {"ijk", matmul_ijk},
        {"ikj", matmul_ikj},
        {"jik", matmul_jik},
        {"jki", matmul_jki},
        {"kij", matmul_kij},
        {"kji", matmul_kji},
    };
    const int NORDERS = 6;

    std::cout << "=========================================================\n";
    std::cout << "  Multiplicación de matrices C++ — orden de bucles\n";
    std::cout << "=========================================================\n\n";

    // Encabezado
    std::cout << std::left << std::setw(6) << "Orden";
    for (int s = 0; s < NSIZES; s++)
        std::cout << std::right << std::setw(12) << ("N=" + std::to_string(SIZES[s]));
    std::cout << "\n" << std::string(6 + NSIZES * 12, '-') << "\n";

    for (int o = 0; o < NORDERS; o++) {
        std::cout << std::left << std::setw(6) << orders[o].name;
        for (int s = 0; s < NSIZES; s++) {
            int N = SIZES[s];
            if (N > 512 && (strcmp(orders[o].name,"kji")==0 || strcmp(orders[o].name,"jki")==0)) {
                // estos órdenes son muy lentos para N=1024 en demo rápido
                init(N);
                double ms = measure(orders[o].fn, N, 1);
                std::cout << std::right << std::setw(10) << std::fixed << std::setprecision(1) << ms << "ms";
            } else {
                init(N);
                double ms = measure(orders[o].fn, N);
                std::cout << std::right << std::setw(10) << std::fixed << std::setprecision(1) << ms << "ms";
            }
        }
        std::cout << "\n";
    }

    std::cout << "\n";
    std::cout << "Leyenda de acceso a caché:\n";
    std::cout << "  ikj / kij — stride-1 en j (inner loop)  → MEJOR\n";
    std::cout << "  ijk / jik — stride-N en B[k][j]         → MEDIO\n";
    std::cout << "  jki / kji — stride-N en A[i][k] y C[i][j] → PEOR\n";
    std::cout << "\nComplejidad: O(N³) = " << (long long)1024*1024*1024 << " ops para N=1024\n";
    return 0;
}

File: TAREA1/3_bucles_anidados.go
package main

import (
	"fmt"
	"math/rand"
	"strings"
	"time"
)

const MAXN = 1024

// Matrices como arrays planos (row-major) para garantizar contigüidad en memoria
// igual que en C/C++. A[i][j] = data[i*N + j]
type Matrix struct {
	data [MAXN * MAXN]float64
	n    int
}

func (m *Matrix) get(i, j int) float64        { return m.data[i*m.n+j] }
func (m *Matrix) set(i, j int, v float64)     { m.data[i*m.n+j] = v }
func (m *Matrix) add(i, j int, v float64)     { m.data[i*m.n+j] += v }

var A, B, C Matrix

func init_matrices(N int) {
	rng := rand.New(rand.NewSource(42))
	A.n, B.n, C.n = N, N, N
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			A.set(i, j, rng.Float64())
			B.set(i, j, rng.Float64())
		}
	}
}

func zero_C(N int) {
	for i := 0; i < N*N; i++ {
		C.data[i] = 0
	}
}

// ─── Los 6 órdenes de bucle ──────────────────────────────────────

// ijk — clásico. B[k][j] con stride-N en inner loop → medio
func matmul_ijk(N int) {
	zero_C(N)
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			for k := 0; k < N; k++ {
				C.add(i, j, A.get(i, k)*B.get(k, j))
			}
		}
	}
}

// ikj — MEJOR orden. j varía en inner loop: A[i][k] escalar,
//        B[k][j] y C[i][j] ambos stride-1 ✓
func matmul_ikj(N int) {
	zero_C(N)
	for i := 0; i < N; i++ {
		for k := 0; k < N; k++ {
			r := A.get(i, k) // escalar: sólo 1 carga de memoria
			base_b := k * N
			base_c := i * N
			for j := 0; j < N; j++ {
				C.data[base_c+j] += r * B.data[base_b+j] // stride-1 ✓
			}
		}
	}
}

// jik
func matmul_jik(N int) {
	zero_C(N)
	for j := 0; j < N; j++ {
		for i := 0; i < N; i++ {
			for k := 0; k < N; k++ {
				C.add(i, j, A.get(i, k)*B.get(k, j))
			}
		}
	}
}

// jki
func matmul_jki(N int) {
	zero_C(N)
	for j := 0; j < N; j++ {
		for k := 0; k < N; k++ {
			r := B.get(k, j)
			for i := 0; i < N; i++ {
				C.add(i, j, A.get(i, k)*r)
			}
		}
	}
}

// kij — también muy bueno, similar a ikj
func matmul_kij(N int) {
	zero_C(N)
	for k := 0; k < N; k++ {
		for i := 0; i < N; i++ {
			r := A.get(i, k)
			base_b := k * N
			base_c := i * N
			for j := 0; j < N; j++ {
				C.data[base_c+j] += r * B.data[base_b+j]
			}
		}
	}
}

// kji — PEOR. Ambos A y C tienen stride-N en el inner loop
func matmul_kji(N int) {
	zero_C(N)
	for k := 0; k < N; k++ {
		for j := 0; j < N; j++ {
			r := B.get(k, j)
			for i := 0; i < N; i++ {
				C.add(i, j, A.get(i, k)*r)
			}
		}
	}
}

// ─── Benchmark ───────────────────────────────────────────────────
type MatmulFn func(int)

func measure(fn MatmulFn, N, runs int) float64 {
	var total time.Duration
	for r := 0; r < runs; r++ {
		init_matrices(N)
		start := time.Now()
		fn(N)
		total += time.Since(start)
	}
	return float64(total.Milliseconds()) / float64(runs)
}

func main() {
	sizes := []int{64, 128, 256, 512, 1024}

	type Order struct {
		name string
		fn   MatmulFn
	}
	orders := []Order{
		{"ijk", matmul_ijk},
		{"ikj", matmul_ikj},
		{"jik", matmul_jik},
		{"jki", matmul_jki},
		{"kij", matmul_kij},
		{"kji", matmul_kji},
	}

	fmt.Println("=========================================================")
	fmt.Println("  Multiplicación de matrices Go — orden de bucles")
	fmt.Println("=========================================================\n")

	// Encabezado
	fmt.Printf("%-6s", "Orden")
	for _, n := range sizes {
		fmt.Printf("%12s", fmt.Sprintf("N=%d", n))
	}
	fmt.Println()
	fmt.Println(strings.Repeat("-", 6+len(sizes)*12))

	for _, ord := range orders {
		fmt.Printf("%-6s", ord.name)
		for _, N := range sizes {
			runs := 3
			if N == 1024 && (ord.name == "jki" || ord.name == "kji") {
				runs = 1 // muy lento, 1 corrida basta para demo
			}
			ms := measure(ord.fn, N, runs)
			fmt.Printf("%10.1fms", ms)
		}
		fmt.Println()
	}

	fmt.Println()
	fmt.Println("Leyenda de acceso a caché:")
	fmt.Println("  ikj / kij — stride-1 en j (inner loop)     → MEJOR")
	fmt.Println("  ijk / jik — stride-N en B[k][j]            → MEDIO")
	fmt.Println("  jki / kji — stride-N en A y C inner loop   → PEOR")
	fmt.Printf("\nComplejidad: O(N³) = %d ops para N=1024\n", 1024*1024*1024)
}
